### DISCUSSION TITLE
SSR Exchange 'cache-and-network'

### DISCUSSION THREAD

**Discussion by yoshiwarab** [urql-graphql/urql#1846] [Q&A] [answered]
Hi I'm using the the ssr exchange with next-urql to hydrate SSR data client-side as described [here](https://formidable.com/open-source/urql/docs/advanced/server-side-rendering/#using-getstaticprops-or-getserversideprops). I have set the request policy within my useQuery call to be 'cache-and-network' however, I think when the data is rehydrated through the ssr exchange, it does not register as a cache hit so no network call is ever made. Is this the intended behavior and if so is there anyway around this without manually re-executing the query?

**Comment by JoviDeCroock** [comment, marked as answer]
Yes, the time difference between ssr rehydration is very small hence why we don't do that currently. Is there a case for this?

EDIT: The fix was done in a PR #1852, we've now introduced the option of specifying `staleWhileRevalidate` as an option

**Reply by yoshiwarab** [reply]
Yes, I'm using [ISR](https://vercel.com/docs/next.js/incremental-static-regeneration) with Next.js so pages are statically generated using public data.  On the client side I have auth exchange set up so I can make authenticated requests. Anyone visiting the site is initially served the statically generated page with public data but another request should be made in the background with the current user's auth so they can see data that is private to them.  This way most people get a super quick page load of the static data, while authenticated users with special access to certain data can still see it. Let me know if that makes sense and thanks for an amazing library!

**Reply by JoviDeCroock** [reply]
Wouldn't that make a better case for ssr then? 😅 the same query returning different data for user makes me go for ssr

**Reply by kitten** [reply]
I think it'd be reasonable to allow the `ssrExchange` to revalidate data. However, if we'd do that I'd explicitly wouldn't make it require `cache-and-network` but instead add a proper "stale-while-revalidate" path to it, since that'd avoid a default that'd refetch excessively

**Reply by yoshiwarab** [reply]
> Wouldn't that make a better case for ssr then? 😅 the same query returning different data for user makes me go for ssr

Yes, but for our use case most people will get the same data so the faster page load of a static page is worth it for us. I think a more common use case though is refetching stale data after loading a statically generated page.  With Next.js a statically generated page isn't re-rendered until after someone triggers the re-render by visiting after a specified `revalidate` period (https://vercel.com/docs/next.js/incremental-static-regeneration#fetching-data). The initial data the user sees, might be stale so it makes sense to re-execute the query client side.

**Reply by yoshiwarab** [reply]
> I think it'd be reasonable to allow the `ssrExchange` to revalidate data. However, if we'd do that I'd explicitly wouldn't make it require `cache-and-network` but instead add a proper "stale-while-revalidate" path to it, since that'd avoid a default that'd refetch excessively

Oh, what would that look like? SSG data is given an expiration time and then only refetched on the client if the time has passed?

**Reply by kitten** [reply]
> Oh, what would that look like? SSG data is given an expiration time and then only refetched on the client if the time has passed?

Exactly, I think that'd actually be quite a nice built-in feature.

**Reply by JoviDeCroock** [reply]
Hmm, would next add some type of timestamp or would we do so? I assume a timestamp on the original entry in this JSON map is what you're implying 😅

**Reply by JoviDeCroock** [reply]
Will be solved by https://github.com/FormidableLabs/urql/pull/1852

**Reply by yoshiwarab** [reply]
Amazing, thank you!
